### PR TITLE
e2e tests: added tests to validate e2e streaming workflow between B and O

### PR DESCRIPTION
**Pull request by RiccardoBiosas**
**What does this pull request do? Explain your changes. (required)**
Adds the test cases to validate the e2e streaming workflow between a broadcaster and an orchestrator - from setting up the orchestrator and depositing the required on-chain broadcaster's funds to the probabilistic micropayment workflow.

**Specific updates (required)**
Updated the files under `tests/e2e`

**How did you test each of these updates (required)**
Run the `test_e2e.sh` script

**Does this pull request close any open issues?**
[](https://github.com/livepeer/go-livepeer/issues/2319)

**Checklist:**

- [ ] Read the [contribution guide](./doc/contributing.md)
- [ ] `make` runs successfully
- [ ] All tests in `./test.sh` pass
- [ ] README and other documentation updated
- [ ] [Pending changelog](./CHANGELOG_PENDING.md) updated

### CODE REVIEW

File: test/e2e/e2e.go [hunk 1]
```diff
@@ -73,11 +73,11 @@ type livepeer struct {
 
 func lpCfg() starter.LivepeerConfig {
 	mu.Lock()
-	serviceAddr := fmt.Sprintf("127.0.0.1:%d", httpPort)
+	serviceAddr := fmt.Sprintf("localhost:%d", httpPort)
```
Comment (leszko) [inline test/e2e/e2e.go:76]: Just out of curiosity, why did you change it? Did `127.0.0.1` cause any issues?

File: test/e2e/e2e.go [hunk 2]
```diff
@@ -202,3 +202,15 @@ func httpPostWithParamsHeaders(url string, val url.Values, headers map[string]st
 
 	return string(result), resp.StatusCode >= 200 && resp.StatusCode < 300
 }
+
+func httpGet(url string) string {
```
Comment (leszko) [inline test/e2e/e2e.go:206]: Why do you use this function? If unused, please remove.

File: test/e2e/register_orchestrator_test.go [hunk 3]
```diff
@@ -30,8 +30,13 @@ func TestRegisterOrchestrator(t *testing.T) {
 
 func startOrchestrator(t *testing.T, geth *gethContainer) *livepeer {
```
Comment (leszko) [inline test/e2e/register_orchestrator_test.go:31]: If this function is used both in `register_orchestrator_test.go` and `streaming_e2e_test.go`, then how about moving it to `e2e.go`?

File: test/e2e/register_orchestrator_test.go [hunk 4]
```diff
@@ -30,8 +30,13 @@ func TestRegisterOrchestrator(t *testing.T) {
...
 	lpCfg := lpCfg()
+	MaxFaceValue := big.NewInt(1000000).String()
+	TicketEV := big.NewInt(1000000).String()
 	lpCfg.Orchestrator = boolPointer(true)
 	lpCfg.Transcoder = boolPointer(true)
+	lpCfg.Redeemer = boolPointer(true)
```
Comment (leszko) [inline test/e2e/register_orchestrator_test.go:37]: I think you don't need to set it to `true`. By default O should redeem the received tickets. AFAIK `redeemer=true` is only used when you run a separate Redeemer service.

File: test/e2e/streaming_e2e_test.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+	"github.com/livepeer/go-livepeer/eth"
+	lpTypes "github.com/livepeer/go-livepeer/eth/types"
+	"github.com/livepeer/go-livepeer/pm"
+	"github.com/stretchr/testify/require"
+)
+
+func TestE2EStreamingWorkflow(t *testing.T) {
+	geth := setupGeth(t)
+	defer terminateGeth(t, geth)
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:22]:
```suggestion
    // given
	geth := setupGeth(t)
	defer terminateGeth(t, geth)
```

File: test/e2e/streaming_e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+	// 1. Start and register Orchestrator
+	o := startOrchestrator(t, geth)
+	// 2. Start Broadcaster
+	b := startBroadcaster(t, geth)
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:27]:
```suggestion
	o := startOrchestrator(t, geth)
	b := startBroadcaster(t, geth)
```

nit: I think you don't need these comments, because the code actually says the same. It's clear that you start both O and B in these lines.

File: test/e2e/streaming_e2e_test.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+	oEth := o.dev.Client
+	bEth := b.dev.Client
+	defer o.stop()
+	defer b.stop()
+	<-o.ready
+	<-b.ready
+
+	// when
+	// 1. Start and register Orchestrator
+	registerOrchestrator(o)
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:37]:
```suggestion
	registerOrchestrator(o)
```

nit: the same here, this comment actually is even misleading because it says `Start`, but you're not starting anything here.

File: test/e2e/streaming_e2e_test.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	waitForNextRound(t, oEth)
+
+	// then
+	fmt.Println("b.Account() orch ", oEth.Account())
+	fmt.Println("b.Account() broadcaster ", bEth.Account())
+	assertOrchestratorRegisteredAndActivated(t, oEth)
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:43]:
```suggestion
	assertOrchestratorRegisteredAndActivated(t, oEth)
```

nit: I'd avoid printlns in the test.

File: test/e2e/streaming_e2e_test.go [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+	// when
+	// 3. Fund Deposit/Reserve
+	fundBroadcasterDepositAndReserve(b)
+
+	// then
+	// 4. Check if broadcaster's funds have been correctly deposited
+	assertBroadcasterFundsDeposited(t, b, bEth)
+	// // waitForNextRound(t, bEth)
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:52]: Please clean up the comments.

File: test/e2e/streaming_e2e_test.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+	}
+
+	if _, ok := httpPostWithParams(fmt.Sprintf("http://%s/fundDepositAndReserve", *b.cfg.CliAddr), val); ok {
+		return
+	}
+	time.Sleep(200 * time.Millisecond)
+
+}
+
+func getBroadcasterFunds(urlAddr string, b eth.LivepeerEthClient) pm.SenderInfo {
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:98]: 1. I think the `b` parameter is not used. If that's the case, how about removing it?
2. Nit: Consider moving this function under `assertBroadcasterFundsDeposited`. In general, it's easier to read code, when the used function is just under the using one.

File: test/e2e/streaming_e2e_test.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+	require := require.New(t)
+
+	broadcasterFunds := getBroadcasterFunds(*bNode.cfg.CliAddr, b)
+
+	require.Equal(broadcasterFunds.Deposit, big.NewInt(1000000))
+	require.Equal(broadcasterFunds.Reserve.FundsRemaining, big.NewInt(1000000))
+}
+
+func startStream(b *livepeer) {
+	fakeSegment, _ := ioutil.ReadFile("../../server/test.ts")
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:117]: Nit: Consider copying this `test.ts` into the `test/e2e` directory. I think E2E Test can have their own testing segments.

File: test/e2e/streaming_e2e_test.go [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+	reader := bytes.NewReader(fakeSegment)
+
+	response, _ := http.Post(fmt.Sprintf("http://%s/live/name/12.ts", *b.cfg.ServiceAddr), "application/x-www-form-urlencoded", reader)
+
+	defer response.Body.Close()
+
+}
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:124]:
```suggestion
	defer response.Body.Close()
}
```

File: test/e2e/streaming_e2e_test.go [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+
+func assertSuccessfulStreamTranscoding(t *testing.T, b *livepeer) {
+	require := require.New(t)
+	response, _ := http.Get(fmt.Sprintf("http://%s/stream/current.m3u8", *b.cfg.ServiceAddr))
+	defer response.Body.Close()
+	body, _ := ioutil.ReadAll(response.Body)
+
+	require.Equal(200, response.StatusCode)
+	require.Equal("#EXTM3U\n#EXT-X-VERSION:3\n#EXT-X-STREAM-INF:PROGRAM-ID=0,BANDWIDTH=4000000,RESOLUTION=0x0\nname/source.m3u8\n", string(body))
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:133]: I wonder, won't this header string never change? If it's possible that this header changes in the future, maybe we could be a little less strict here and check something like this?

```
require.Contains(string(body), "source.m3u8")
```

File: test/e2e/streaming_e2e_test.go [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+	delegateInfoResponseBefore, _ := http.Get(fmt.Sprintf("http://%s/delegatorInfo", *o.cfg.CliAddr))
+	delegateInfoResponseBody, _ := ioutil.ReadAll(delegateInfoResponseBefore.Body)
+	json.Unmarshal(delegateInfoResponseBody, &delegatorInfoBefore)
+
+	httpPostWithParams(fmt.Sprintf("http://%s/claimEarnings", *o.cfg.CliAddr), nil)
+
+	delegateInfoResponseAfter, _ := http.Get(fmt.Sprintf("http://%s/delegatorInfo", *o.cfg.CliAddr))
+	delegateInfoBodyAfter, _ := ioutil.ReadAll(delegateInfoResponseAfter.Body)
+	json.Unmarshal(delegateInfoBodyAfter, &delegatorInfoAfter)
+	require.GreaterOrEqual(delegatorInfoAfter.Fees.Uint64(), delegatorInfoBefore.Fees.Uint64())
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:151]: I think that `GreaterOrEqual` does not really check what we need, because even if O didn't receive any payment, it will pass the test. Wdyt?

File: test/e2e/streaming_e2e_test.go [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+	delegateInfoResponseAfter, _ := http.Get(fmt.Sprintf("http://%s/delegatorInfo", *o.cfg.CliAddr))
+	delegateInfoBodyAfter, _ := ioutil.ReadAll(delegateInfoResponseAfter.Body)
+	json.Unmarshal(delegateInfoBodyAfter, &delegatorInfoAfter)
+
+	oEthBalanceResponseAfter, _ := http.Get(fmt.Sprintf("http://%s/ethBalance", *o.cfg.CliAddr))
+	oEthBalanceResponseBodyAfter, _ := ioutil.ReadAll(oEthBalanceResponseAfter.Body)
+	json.Unmarshal(oEthBalanceResponseBodyAfter, &oEthBalanceAfter)
+
+	require.Equal(delegatorInfoAfter.Fees, big.NewInt(0))
+	require.Greater(oEthBalanceAfter.Uint64(), oEthBalanceBefore.Uint64())
```
Comment (leszko) [inline test/e2e/streaming_e2e_test.go:183]: How do we ensure that the winning ticket is received? Don't we need to set `maxFaceValue` or `winProb` somewhere? Otherwise, if the ticket is probabilistic, how can we ensure that the winning ticket is received?

**Review by leszko — CHANGES_REQUESTED** [review]
Thanks for the PR @RiccardoBiosas . Added some comments.

Other than that, could you also check the following points:
- Update Description => Add `fix #2319` under `Does this pull request close any open issues?` (then merging your PR will automatically close the related GH Issue)
- Update Description => Mark the checklist (either check the box or cross out the whole point if it's not related)
- Please contact with @red-0ne  how he uses the `context`. Otherwise you're test will fail on `error getting gas price: Post "http://localhost:49960": dial tcp [::1]:49960: connect: connection refused`
- GH Action failed ([link](https://github.com/livepeer/go-livepeer/runs/6880494560)), could you check it?
- `./test_e2e.sh` fails for me locally as well, could you check? Here's the error.

```
--- FAIL: TestE2EStreamingWorkflow (112.98s)
    assertion_compare.go:313: 
                Error Trace:    streaming_e2e_test.go:183
                                                        streaming_e2e_test.go:68
                Error:          "3895020945059707114" is not greater than "3895108342060318893"
                Test:           TestE2EStreamingWorkflow
                Messages:       []
FAIL
FAIL    github.com/livepeer/go-livepeer/test/e2e        158.956s
FAIL
```

File: test/e2e/register_orchestrator_test.go (hunk 3 above)
Comment (red-0ne) [reply test/e2e/register_orchestrator_test.go:31]: It would be interesting in general to move all the partial flows to `e2e.go` / dedicated directory. IMO, this will help reuse these instead of heading to write test-specific ones. Our test-writing-flow should involve looking for and using appropriate partial-flow from the ones already there, **then** if none is suitable, try to add params to close/similar existing ones or; if it's not possible write a new one.

File: test/e2e/streaming_e2e_test.go [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+	defer b.stop()
+	<-o.ready
+	<-b.ready
+
+	// when
+	// 1. Start and register Orchestrator
+	registerOrchestrator(o)
+	waitForNextRound(t, oEth)
+
+	// then
```
Comment (red-0ne) [inline test/e2e/streaming_e2e_test.go:40]: This seems like a "nested" then, I started using the same pattern but it's kinda tricky to express/signify these nested thens.
I think `//then` should be for the final outcome and intermediary assertions like `assertOrchestratorRegisteredAndActivated` are just intermediary checks used to fail tests early.

```
// given
setupTest() // everything needed to setup the test

// when
doSomeAction1()
requireSomeConditionToBeMet() // prefer requireXXX over assertXXX for non final checks
doSomeAction2()

// then
assertFinalState() // this is the final outcome we are looking for
```

This is only an esthetic/dev experience change that may help better reason and maintain e2e tests. WDYT @RiccardoBiosas, @leszko ?

File: test/e2e/streaming_e2e_test.go [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+	json.Unmarshal(delegateInfoResponseBody, &delegatorInfoBefore)
+
+	oEthBalanceResponseBefore, _ := http.Get(fmt.Sprintf("http://%s/ethBalance", *o.cfg.CliAddr))
+	oEthBalanceResponseBodyBefore, _ := ioutil.ReadAll(oEthBalanceResponseBefore.Body)
+	json.Unmarshal(oEthBalanceResponseBodyBefore, &oEthBalanceBefore)
+	val := url.Values{
+		"amount": {fmt.Sprintf("%v", delegatorInfoBefore.PendingFees.String())},
+	}
+
+	httpPostWithParams(fmt.Sprintf("http://%s/withdrawFees", *o.cfg.CliAddr), val)
```
Comment (red-0ne) [inline test/e2e/streaming_e2e_test.go:172]: I don't think we should alter state (do post requests) inside assertion functions.
What about
```
withdrawFunds(...)
assertWithdrawalSuccessful(...)
```
I know it's tricky because you want to compare old an new values but we can do something like
```
oldData := getOrchestratorBalanceAndInfo(o)
withdrawFunds(o)
assertWithdrawalSuccessful(o, oldData)
```

Does this makes sense?

File: test/e2e/streaming_e2e_test.go (hunk 15 above)
Comment (red-0ne) [reply test/e2e/streaming_e2e_test.go:183]: > ... Don't we need to set `maxFaceValue` or `winProb` somewhere?

isn't this enough?
`test/e2e/register_orchestrator_test.go#L33`
```go
...
MaxFaceValue := big.NewInt(1000000).String()
...
```

File: test/e2e/streaming_e2e_test.go [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+func assertEarnings(t *testing.T, o *livepeer) {
+	require := require.New(t)
+	var delegatorInfoBefore lpTypes.Delegator
+	var delegatorInfoAfter lpTypes.Delegator
+
+	delegateInfoResponseBefore, _ := http.Get(fmt.Sprintf("http://%s/delegatorInfo", *o.cfg.CliAddr))
+	delegateInfoResponseBody, _ := ioutil.ReadAll(delegateInfoResponseBefore.Body)
+	json.Unmarshal(delegateInfoResponseBody, &delegatorInfoBefore)
+
+	httpPostWithParams(fmt.Sprintf("http://%s/claimEarnings", *o.cfg.CliAddr), nil)
```
Comment (red-0ne) [inline test/e2e/streaming_e2e_test.go:146]: Same as for withdrawal below, try to avoid changing state inside assertions.
